### PR TITLE
CADC-10640: cadc-vos-server and cadc-test-vos, cavern using cadc-vos 2.0

### PR DESCRIPTION
**Pull request by hjeeves**
…o new Transfer.java in cadc-vos 2.0

### CODE REVIEW

**Review by brianmajor — CHANGES_REQUESTED** [review]
It's true that BiDirectionalTransferNegotiation will never support zip/tar -- it represents an open cavern mount I believe.

PushToVoSpaceTransferNegotiation is an upload so it's correct that multiple targets are not supported there.

PullFromVoSpaceTransferNegotiation will however be represented in new transfer docs with multiple targets.  So perhaps you comments that are in TransferNegotiation.java below in PullFromVoSpaceTransferNegotiation?

I'm not sure if the code in PullFromVoSpaceTransferNegotiation will actually be execute when view=zip/tar.  It could be that the view makes TransferRunner skip some of these steps.  TBD.

File: cadc-test-vos/src/main/java/ca/nrc/cadc/conformance/vos/AsyncPullFromVOSpaceTest.java [hunk 1]
```diff
@@ -142,7 +142,9 @@ public void testPullFromVOSpace()
             p.setSecurityMethod(Standards.SECURITY_METHOD_CERT);
             protocols.add(p);
             
-            Transfer transfer = new Transfer(testNode.sampleNode.getUri().getURI(), Direction.pullFromVoSpace, view, protocols);
+            Transfer transfer = new Transfer(testNode.sampleNode.getUri().getURI(), Direction.pullFromVoSpace);
+            transfer.getProtocols().addAll(protocols);
+            transfer.setView(view);
```
[the remark this answers is not in the document]
Comment (hjeeves) [reply cadc-test-vos/src/main/java/ca/nrc/cadc/conformance/vos/AsyncPullFromVOSpaceTest.java:147]: It must not have picked up cadc-vos 2.0.  That ctor doesn't exist anymore.

File: cadc-vos-server/src/main/java/ca/nrc/cadc/vos/server/transfers/TransferUtil.java [hunk 2]
```diff
@@ -284,5 +291,22 @@ public static String getNamespace(final VOSURI vosURI)
         return ns;
     }
 
+    /**
+     * Check target list size in transfer provided.
+     * @param transfer
+     * @throws TransferException
+     */
+    public static void confirmSingleTarget(Transfer transfer) throws TransferException {
```
[the remark this answers is not in the document]
Comment (hjeeves) [reply cadc-vos-server/src/main/java/ca/nrc/cadc/vos/server/transfers/TransferUtil.java:299]: It's used in PushFromVOSpaceNegotiation, PullFromVOSpaceNegotiation and BiDirectionalTransfer.

File: cadc-vos-server/src/main/java/ca/nrc/cadc/vos/server/transfers/TransferRunner.java [hunk 3]
```diff
@@ -324,6 +324,9 @@ else if (direction.equals(Direction.BIDIRECTIONAL))
             }
             catch (NodeNotFoundException ignore)
```
[the remark this answers is not in the document]
Comment (hjeeves) [reply cadc-vos-server/src/main/java/ca/nrc/cadc/vos/server/transfers/TransferRunner.java]: done.

File: cadc-vos-server/src/main/java/ca/nrc/cadc/vos/server/transfers/TransferRunner.java [hunk 4]
```diff
@@ -324,6 +324,9 @@ else if (direction.equals(Direction.BIDIRECTIONAL))
...
             {
+                // Assertion: if the code gets to here there will be at least one
+                // target available to access
+                VOSURI vosURI = new VOSURI(transfer.getTargets().get(0));
```
[the remark this answers is not in the document]
Comment (hjeeves) [reply cadc-vos-server/src/main/java/ca/nrc/cadc/vos/server/transfers/TransferRunner.java]: removed, made this log.debug statement similar to the ones in the other catch stanzas.

File: cadc-vos-server/src/main/java/ca/nrc/cadc/vos/server/transfers/TransferDetailsServlet.java [hunk 5]
```diff
@@ -309,7 +320,9 @@ public Object run() throws Exception
                         if (!transfer.getProtocols().isEmpty()) {
                             proto = TransferUtil.getTransferEndpoints(transfer, job, additionalParams);
                         }
-                        Transfer result = new Transfer(transfer.getTarget(), dir, proto);
+                        // This is save for now because of the check above (CADC-10640)
```
[the remark this answers is not in the document]
Comment (hjeeves) [reply cadc-vos-server/src/main/java/ca/nrc/cadc/vos/server/transfers/TransferDetailsServlet.java]: done

File: cadc-vos-server/build.gradle [hunk 6]
```diff
@@ -18,7 +18,7 @@ sourceCompatibility = 1.8
 
 group = 'org.opencadc'
 
-version = '1.2.3'
+version = '1.2.4'
```
[the remark this answers is not in the document]
Comment (hjeeves) [reply cadc-vos-server/build.gradle]: changed to 1.3

File: cadc-test-vos/build.gradle [hunk 7]
```diff
@@ -25,7 +25,7 @@ dependencies {
     compile 'org.json:json:20160212'
 
     compile 'org.opencadc:cadc-util:[1.3,2.0)'
-    compile 'org.opencadc:cadc-vos:[1.0,2.0)'
+    compile 'org.opencadc:cadc-vos:[2.0,)'
     compile 'org.opencadc:cadc-uws:[1.0,2.0)'
     compile 'org.opencadc:cadc-registry:[1.5,2.0)'
```
[the remark this answers is not in the document]
Comment (hjeeves) [reply cadc-test-vos/build.gradle:31]: done

**Review by brianmajor — APPROVED** [review]

**Review by brianmajor — CHANGES_REQUESTED** [review]
Null pointer in TransferReader must be fixed first.  Seen when running the cavern int tests.

```	
java.lang.NullPointerException
at java.util.ArrayList.addAll(ArrayList.java:581)
at ca.nrc.cadc.vos.TransferReader.parseTransfer(TransferReader.java:213)
at ca.nrc.cadc.vos.TransferReader.read(TransferReader.java:146)
at ca.nrc.cadc.vos.TransferReader.read(TransferReader.java:157)
at ca.nrc.cadc.vos.server.TransferInlineContentHandler.accept(TransferInlineContentHandler.java:120)
	```

**Review by brianmajor — APPROVED** [review]
cavern int-tests now passing with the update

**Pull request merged**